File: cpp/1305_AllElementsInTwoBinarySearchTrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> first;
    vector<int> second;
    vector<int> getAllElements(TreeNode* root1, TreeNode* root2) {
        inorder_dfs(root1, first);
        inorder_dfs(root2, second);
        
        return combine(first, second);
    }
    
    void inorder_dfs(TreeNode* node, vector<int> &arr) {
        if (!node) 
            return;
        
        if (node->left)
            inorder_dfs(node->left, arr);
        
        arr.push_back(node->val);
        
        if (node->right)
            inorder_dfs(node->right, arr);
    }
    
    vector<int> combine(vector<int> &a, vector<int> &b) {
        int left = 0, right = 0;
        vector<int> res;
        while (left < a.size() || right < b.size()) {
            if (left < a.size() && right < b.size()) {
                if (a[left] < b[right])
                    res.push_back(a[left++]);
                else 
                    res.push_back(b[right++]);
            } else if (right < b.size()) {
                res.push_back(b[right++]);
            } else {
                res.push_back(a[left++]);
            }
        }
        return res;
    }
};

File: swift/1305_AllElementsInTwoBinarySearchTrees.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func getAllElements(_ r1: TreeNode?, _ r2: TreeNode?) -> [Int] {
        var root1: TreeNode? = r1, root2: TreeNode? = r2
        var stack1: [TreeNode?] = [TreeNode?]()
        var stack2: [TreeNode?] = [TreeNode?]()
        var res: [Int] = [Int]()
        
        while !stack1.isEmpty || !stack2.isEmpty || root1 != nil || root2 != nil {
            // left branches in stack
            while root1 != nil {
                stack1.append(root1)
                root1 = root1?.left
            }
            
            while root2 != nil {
                stack2.append(root2)
                root2 = root2?.left
            }
            
            if stack2.isEmpty || !stack1.isEmpty && stack1[stack1.endIndex-1]!.val <= stack2[stack2.endIndex - 1]!.val {
                root1 = stack1.popLast()!
                res.append(root1!.val)
                root1 = root1?.right
            } else {
                root2 = stack2.popLast()!
                res.append(root2!.val)
                root2 = root2?.right
            }
        }
        return res
    }
}
